enumerable.rb

```diff
@@ -0,0 +1,233 @@
+module Enumerable
+  def my_each
+    return to_enum unless block_given?
+
+    i = 0
+    while i < size
+      yield(to_a[i])
+      i += 1
+    end
+    self
+  end
+
+  def my_each_with_index
+    return to_enum unless block_given?
+
+    i = 0
+    while i < size
+      yield(to_a[i], i)
+      i += 1
+    end
+    self
+  end
+
+  def my_select
+    return to_enum unless block_given?
+
+    result = []
+    my_each do |item|
+      result << item if yield(item)
+    end
+    result
+  end
+
+  def my_all?(param = nil)
+    result = true
+    my_each do |value|
+      if !block_given?
+        result = false unless param === value
+      elsif param.nil?
+        result = true unless value
+      else
+        result = false unless yield(value)
+      end
+    end
+    result
+  end
+
+  def my_any?(param = nil)
+    result = false
+    my_each do |value|
+      if !block_given?
+        result = true if param === value
+      elsif param.nil?
+        result = true if value
+      else
+        result = true if yield(value)
+      end
+    end
+    result
+  end
+
+  def my_none?(param = nil)
+    result = true
+    my_each do |value|
+      if block_given?
+        result = false if yield(value)
+      elsif param.nil?
+        result = false if value
+      else
+        result = false if param === value
+      end
+    end
+    result
+  end
+
+  def my_count(param = nil)
+    count = 0
+    my_each do |item|
+      if param
+        count += 1 if item == param
+      elsif block_given?
+        count += 1 if yield(item)
+      else
+        count = length
+      end
+    end
+    count
+  end
+
+  def my_map(param = nil)
+    return to_enum unless block_given?
```

Comment:
* [ ] Improve this by passing the name of the method as an argument to to_enum(method_name).return to_enum(method_name) unless block_given? instead of return to_enum unless block_given?

Reply:
done